src/libcollections/string.rs

```diff
@@ -701,6 +701,13 @@ impl String {
     /// Violating these may cause problems like corrupting the allocator's
     /// internal datastructures.
     ///
+    /// The ownership of `ptr` is effectively transferred to the
+    /// `String` and the memory referred to by the pointer will be
+    /// deallocated when the `String` goes out of scope. Ensure that
+    /// nothing else uses the pointer after calling this
+    /// function. Calling this function multiple times with the same
+    /// pointer is an error.
```

Comment:
I feel like the “The ownership of `ptr` is effectively transferred to the `String`” part already covers what you’re trying to say here, without claiming invalidity of any particular use of the function. You perhaps could word that sentence this way too:

> The ownership of `ptr` is effectively transferred to the `String`. `String` then may deallocate, reallocate or change the contents of memory pointed to by the pointer at will.

Of course, `mem::forget(String::from_raw_parts(x, ...)); String::from_raw_parts(x, ...)` is something that relies on internal implementation details of `String`, is unstable and may not work in the future. For example, if we begin `realloc`-ing the pointer in question to the next power of two in `from_raw_parts` for some reason. But I still am not sure about calling reliance on implementation detail an error.